feat: a simmple demo for RW as PG-fdw (#2055)
risingwavelabs/risingwave-docs: docs/guides/risingwave-as-postgres-fdw.md

---
id: risingwave-as-postgres-fdw
title: Use RisingWave as a Postgres Foreign Data Wrapper
description: Use RisingWave as a Postgres Foreign Data Wrapper
slug: /risingwave-as-postgres-fdw
---

> Postgres's FDW (Foreign Data Wrapper) allows you to directly virtualize data stored in an external database as a local external table, also known as a foreign table. This tutorial will demonstrate how to interact between Postgres (PostgreSQL) and RisingWave. In this example, RisingWave will use CDC (Change Data Capture) to extract data from Postgres, analyze it using a Materialized View, then Postgres will directly retrieve the computation results stored in RisingWave.

Reviewer: ```suggestion
Postgres's FDW (Foreign Data Wrapper) allows you to directly virtualize data stored in an external database as a local external table, also known as a foreign table. This tutorial will demonstrate how to interact between Postgres (PostgreSQL) and RisingWave. In this example, RisingWave will use CDC (Change Data Capture) to extract data from Postgres, analyze it using a materialized view, then Postgres will directly retrieve the computation results stored in RisingWave.
```

Author: updated.

Author: BTW, should we always use `PostgreSQL`?

:::note Beta Feature

> The RisingWave's support as a Postgres' FDW is currently in Beta. Please contact us if you encounter any issues or have feedback.

Reviewer: ```suggestion
RisingWave serving as a foreigh data wrapper of PostgreSQL is currently in Beta. Please contact us if you encounter any issues or have feedback.
```

Author: updated

INSERT INTO person VALUES (1016, 'john walton', '0426 2682 6145 8371', 'seattle');
INSERT INTO person VALUES (1017, 'luke jones', '9641 9352 0248 2749', 'redmond');
```

## Analyze data in RisingWave

The following command creates a table in RisingWave. This table will use Native CDC to synchronize the data of the Person table from Postgres, and then create a materialized view to analyze the ingested data.

```sql
---Run in RisingWave
---Create a table in RisingWave to replicate the Person table of Postgres into RisingWave
CREATE TABLE pg_person (
    "id" int,
    "name" varchar,
    "credit_card" varchar,
    "city" varchar,
    PRIMARY KEY ("id")
) with (
    connector = 'postgres-cdc',
    hostname = '127.0.0.1',
    port = '5432',
    username = 'postgresuser',
    password = 'postgrespw',
    database.name = 'mydb',
    schema.name = 'public',
    table.name = 'person',
    slot.name = 'person'
);

---Create a materialized view to analyze the population of each city
CREATE MATERIALIZED VIEW city_population AS
SELECT
    city,
    COUNT(*) as population
FROM
    pg_person
GROUP BY
    city;
```

> ## Query result in Postgres using FDW

Reviewer: @KeXiangWang  What's the difference between sinking data to PG and PG pulling data from RW?

Author: added one section for comparison, PTAL

The following command creates a foreign table in Postgres to connect to RisingWave and query the materialized view.

```sql

> ---Run in Postgres

Reviewer: @KeXiangWang Can you specify how to switch from RW to PG if we use psql for both? I guess it'll involve opening a new terminal window and specify the host / port/ dbname / username /password etc.

Author: added, PTAL

  "city" varchar(200),
  PRIMARY KEY ("id")
);

INSERT INTO person VALUES (1001, 'peter white', '1781 2313 8157 6974', 'boise');
INSERT INTO person VALUES (1002, 'sarah spencer', '3453 4987 9481 6270', 'los angeles');
INSERT INTO person VALUES (1004, 'julie white', '0052 8113 1582 4430', 'seattle');
INSERT INTO person VALUES (1005, 'sarah smith', '4591 5419 7260 8350', 'los angeles');
INSERT INTO person VALUES (1007, 'walter spencer', '5136 7504 2879 7886', 'los angeles');
INSERT INTO person VALUES (1008, 'john abrams', '6064 8548 6057 2021', 'redmond');
INSERT INTO person VALUES (1010, 'kate smith', '9474 6887 6463 6972', 'bend');
INSERT INTO person VALUES (1011, 'vicky noris', '9959 4034 5717 6729', 'boise');
INSERT INTO person VALUES (1012, 'walter jones', '8793 6517 3085 0542', 'boise');
INSERT INTO person VALUES (1013, 'sarah walton', '2280 4209 8743 0735', 'kent');
INSERT INTO person VALUES (1015, 'vicky jones', '3148 5012 3225 2870', 'los angeles');
INSERT INTO person VALUES (1016, 'john walton', '0426 2682 6145 8371', 'seattle');
INSERT INTO person VALUES (1017, 'luke jones', '9641 9352 0248 2749', 'redmond');
```

## Analyze data in RisingWave

> The following command creates a table in RisingWave. This table will use Native CDC to synchronize the data of the Person table from PostgreSQL, and then create a materialized view to analyze the ingested data.

Reviewer: ```suggestion
The following command creates a table in RisingWave. This table will use the native CDC connector to synchronize the data of the Person table from PostgreSQL, and then create a materialized view to analyze the ingested data.
```

  id  |      name      |     credit_card     |    city
------+----------------+---------------------+-------------
 1005 | sarah smith    | 4591 5419 7260 8350 | los angeles
 1012 | walter jones   | 8793 6517 3085 0542 | boise
 1002 | sarah spencer  | 3453 4987 9481 6270 | los angeles
 1007 | walter spencer | 5136 7504 2879 7886 | los angeles
 1011 | vicky noris    | 9959 4034 5717 6729 | boise
 1016 | john walton    | 0426 2682 6145 8371 | seattle
 1010 | kate smith     | 9474 6887 6463 6972 | bend
 1015 | vicky jones    | 3148 5012 3225 2870 | los angeles
 1017 | luke jones     | 9641 9352 0248 2749 | redmond
 1001 | peter white    | 1781 2313 8157 6974 | boise
 1004 | julie white    | 0052 8113 1582 4430 | seattle
 1008 | john abrams    | 6064 8548 6057 2021 | redmond
 1013 | sarah walton   | 2280 4209 8743 0735 | kent

---Query the materialized view in RisingWave with PostgreSQL's foreign table.
SELECT * FROM city_population;
-------------+------------
    city     | population
-------------+------------
 boise       |          3
 los angeles |          4
 bend        |          1
 kent        |          1
 redmond     |          2
 seattle     |          2
```

> ## Difference between Sink and FDW

Reviewer: ```suggestion
## Differences between sink and FDW
```

> Both sinking data to PostgreSQL and using a foreigh data wrapper of PostgreSQL to access data in RisingWave are ways to interact between RisingWave and PostgreSQL. The following table summarizes the differences between the two methods. The choice between these methods depends on your specific needs, data architecture, and performance considerations.

Reviewer: ```suggestion
There are two main methods to interact between RisingWave and PostgreSQL: sinking data to PostgreSQL and utilizing a foreign data wrapper of PostgreSQL to access data in RisingWave. The table below provides a summary of the differences between these two methods. Your choice between these methods will depend on your specific requirements, data architecture, and performance considerations.
```

| Aspect            | Sinking to PostgreSQL                     | Using PostgreSQL FDW to access data               |
|-------------------|-------------------------------------------|---------------------------------------------------|
| Data Access       | Data is physically stored in PostgreSQL   | Data is physically stored in RisingWave           |
| Performance       | Potential latency for RisingWave writing to PostgreSQL | Potential latency when reading data from RisingWave |

> | Data Consistency  | High; data written to PostgreSQL is immediately transaction-ready | Data in RisingWave doesn't support read-write transactions |

Reviewer 2: ```suggestion
| Data Consistency  | High; data written to PostgreSQL is immediately transaction-ready | Data in RisingWave doesn't support read-write transactions |
```

I am not sure if this is correct. IIUC, these two methods provide equal levels of data consistency as well as data correctness. @wenym1 may take a look

Reviewer 3: Yes. I think even fdw support better data consistency, because we have exactly-once delivery from upstream source to mv, but only at-least-once from mv and sink, so in some cases there can be even data duplication in the downstream sink (e.g. for append only sink without pk to do upsert)

Author: Updated, PTAL

Reviewer 3: Changes on this part LGTM.

> | Complexity        | Lower; straightforward data pipeline      | Higher; involves managing FDW setup               |

Reviewer 2: TBH, I feel FDW has lower complexity if the read latency is not a concern. I am not sure why do you think managing FDW is more complex?

Author: From the user's perspective, there are two reasons:

- FDW requires the support for the `postgres_fdw` extension of PG and involves some additional setup.
- When there is a new table, sinking to PostgreSQL only requires one SQL statement to create the sink, whereas FDW requires first creating a Materialized View (MV) in RW and then establishing a foreign table mapping in PostgreSQL.

If considering system design and code implementation, sinking to a table is certainly more complex.

Reviewer 2: Thanks. I think this is an articulation issue.

```suggestion
| Complexity        | Lower; straightforward data pipeline      | Higher; requires the `postgres_fdw` extension and 
 involves more setup steps.   |
```

BTW, I did some research on several cloud PG vendors, and seems that some vendors, like Neon, indeed do not allow support for `postgres_fdw`. https://neon.tech/docs/extensions/pg-extensions

Author: Updated, PTAL
